### PR TITLE
Add API spec for FQDN selector

### PR DESCRIPTION
**Pull request by rahulkjoshi**
Issue https://github.com/kubernetes-sigs/network-policy-api/issues/133

Provides the proposed API spec as well as some expected behaviors for both application Pods and ANP implementors to follow.

### CODE REVIEW

File: npeps/npep-133-fqdn-egress-selector.md
Comment (astoycos) [inline npeps/npep-133-fqdn-egress-selector.md:234]: Hrm why not?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (astoycos) [inline npeps/npep-133-fqdn-egress-selector.md:229]: Again I think this could be reallly useful for folks

**Review by astoycos — COMMENTED** [review]
Gave this a quick pass and I'll give it another one this week thanks @rahulkjoshi

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:173]: Added another couple examples.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md]: I removed the more complicated regex matching examples. They aren't really necessary, I just got a little carried away with the matcher 😄 

I've left just *.kubernetes.io and *.*.kubernetes.io. I think that captures all the use-cases. I think the user-story for *.*.kubernetes.io is basically the same as for *.kubernetes.io, but for now `*` is just scoped to one subdomain level. Happy to revisit that.

Comment (npinaeva) [reply npeps/npep-133-fqdn-egress-selector.md]: I guess we need to also change validation to not allow asterisks in random places. 
Also wildcard matching may be expensive, and allowing two-subdomain matches raises potential performance concerns, and also questions like: why only 2 subdomain and not 3? So maybe we could start with 1 wildcard match and then extend it in the future?

Comment (npinaeva) [reply npeps/npep-133-fqdn-egress-selector.md]: it was also brought to my attention, that it may be better to only allow one wildcard label and match >0 labels instead of doing `*.*.domain.com` as per https://www.rfc-editor.org/rfc/rfc1034#section-4.3.3
```
Another way to look at this is that the "*" label always matches at least one whole
label and sometimes more, but always whole labels.
```

File: npeps/npep-133-fqdn-egress-selector.md
Comment (Dyanngg) [inline npeps/npep-133-fqdn-egress-selector.md:340]: Thanks for the clarification. In our implementation I don't think we really looked into this and always assumed cname chasing

**Review by Dyanngg — APPROVED** [review]
LGTM

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md]: I don't really have a strong preference one way or the other -- whether `*` matches 1 or multiple labels.

@Dyanngg @astoycos any thoughts?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (astoycos) [inline npeps/npep-133-fqdn-egress-selector.md:101]: This is possible now that the egress npep has merged right, or am I mis-understanding?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (astoycos) [inline npeps/npep-133-fqdn-egress-selector.md:291]: Wait in the "non-goals" section above don't we say 

```
Using the FQDN selector to refer to other Kubernetes endpoints, while not
    explicitly disallowed, is not defined by this spec and left up to individual
    providers. Trying to allow traffic to the following domains is NOT
    guaranteed to work:
    * `my-svc.my-namespace.svc.cluster.local` (the generated DNS record for a
      Service as defined
      [here](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#services))
    * `pod-ip-address.my-namespace.pod.cluster.local` (the generated DNS record
      for a Pod as defined
      [here](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pods))
``` 

So why is this "supported' here?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (astoycos) [reply npeps/npep-133-fqdn-egress-selector.md]: Hrm so I think one of these options
- Leave it up to the implementation to specify
OR
- Start with specifying that it matches only 1 label and expand in the future

(I'd probably go with the latter for now)

**Comment by astoycos** [comment]
Just a few minor questions, otherwise great job @rahulkjoshi!

File: npeps/npep-133-fqdn-egress-selector.md
Comment (npinaeva) [reply npeps/npep-133-fqdn-egress-selector.md]: Starting simple probably makes sense, but we should add a criteria for beta-graduation, saying that we need to revisit the semantics of `*`, since changing it is not backwards-compatible.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (aojea) [inline npeps/npep-133-fqdn-egress-selector.md:138]: is this not a subset of Pattern?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (aojea) [inline npeps/npep-133-fqdn-egress-selector.md:264]: do you want to define some behavior for PR records?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (aojea) [inline npeps/npep-133-fqdn-egress-selector.md:182]: is `host.subdomain.my-service.com` a match?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md]: Ok for now, it matches a single level -- we'll leave it that way. We can revisit upon graduation.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:291]: Whoops, good catch. This was the way I originally wanted to frame it. 
But given the way we defined the network API, I think the discussion in "expected behavior" is more correct.

I'm just removing the section from non-goals.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:101]: I think the idea we wanted to capture is that we need FQDN Selectors at the namespaced NetworkPolicy layer. This was a suggestion from @danwinship to basically remind ourselves that some sort of FQDN selector will eventually end up in NetPolV2 and we don't want to select any behavior here which can't be implemented in a future implementation.

I'm not sure this user-story does a great job of conveying that... I'll try to think of a better phrasing, but suggestions very much welcome.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:138]: It is. The reason I split them is we had discussions originally about core vs extended functionality. There was a concern from folks that some NetPol implementations couldn't support wildcards, but could support a direct match.

Splitting the fields gives those implementations a good way to indicate what they support (they can reject policies which set `pattern`). I'm not sure if we still want that approach

@astoycos do you have thoughts? @tssurya as well since I think you did some work with core vs extended definitions

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:264]: I guess we can say they're not supported, but I imagine they're not very useful in this context?

Like if a customer goes from `1.2.3.4` to `my-hostname`, I don't know what we're even allowlisting here. If they need to allow IPs, the network selector lets you specify CIDR blocks.

Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:264]: Oh I think I understand what you mean. I'm adding a section saying that PTR records aren't required to configure the selector. As long as you get an A record, traffic will flow.

If the PTR is missing or even different (e.g. it points to some other domain), the FQDN selector should still work.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:178]: It's not, the `name` field is supposed to be an exact match. I added some examples to clarify in the API docs.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (npinaeva) [reply npeps/npep-133-fqdn-egress-selector.md:134]: If we want to have separate fields for wildcard and non-wildcard matches sounds like a good section for the NPEP. Let's discuss it during the next meeting and write it down for future reference

File: npeps/npep-133-fqdn-egress-selector.md
Comment (npinaeva) [inline npeps/npep-133-fqdn-egress-selector.md:133]: I have a suggestions around regex:
`([a-zA-z0-9]|[a-zA-Z0-9][-a-zA-Z0-9_]*[a-zA-Z0-9])` can be replaced with 
`[a-zA-z0-9]([-a-zA-Z0-9_]*[a-zA-Z0-9])?`, do you think it is easier to read or does it miss some valid matches?
I am not sure I understand the last `*` in the expression, I think it can be just removed, otherwise `.` also matches.
This expression matches one-label names, like `com`, do we want to enforce at least 2-label match?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (aojea) [reply npeps/npep-133-fqdn-egress-selector.md:134]: it feels weird honestly having both, and this make necessary to treat this as an union, complicating validations, both can not be set at same time

File: npeps/npep-133-fqdn-egress-selector.md
Comment (fasaxc) [inline npeps/npep-133-fqdn-egress-selector.md:154]: Calico (Enterprise) has DNS policy.  Some thoughts from our impl:

* We only have one field for domain vs pattern so that'd be my preference.
* We only support a single wildcard in the domain but it may match more than one dot-separated part.  
* I think we disallow a redundant trailing `.`.  I think that trailing dot has a subtle meaning in some contexts but since we're not honouring that subtle meaning here I say avoid redundancy.

Our wildcarding works like this:

- *.google.com matches www.google.com and www.ipv6.google.com, but not google.com
- www.*.com matches www.sun.com and www.apple.com, but not www.com
- www.google.* matches www.google.net, www.google.io, and so on
- update.*.mycompany.com matches update.tools.mycompany.com, update.secure.suite.mycompany.com, and so on

We don't allow 

- Multiple wildcards in the same domain, for example: *.*.mycompany.com
- Asterisks that are not the entire component, for example: www.g*.com
- More general wildcards, such as regular expressions

I think those limitations were mainly UI driven but the limit of * only matching a whole segment might allow for more efficient indexing (as opposed to needing to do a regex).

File: npeps/npep-133-fqdn-egress-selector.md
Comment (fasaxc) [inline npeps/npep-133-fqdn-egress-selector.md:114]: FQDN feels a bit terse for a k8s API.  Do we lose much by calling this "Domain" or "DomainName" instead?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (fasaxc) [inline npeps/npep-133-fqdn-egress-selector.md:306]: Just to add another gotcha: we had a user who used DNS load balancing to load balance to a set of external IPs, each in a separate k8s cluster.  They then tried to use DNS policy with _that_; it would work when a pod connected to another cluster but not the local cluster because kube-proxy was short-circuiting the load balancer and handling the external IP in-cluster.  

User thought they were using an external load balancer but that load balancer used external IPs under the covers. Very confusing!

File: npeps/npep-133-fqdn-egress-selector.md
Comment (fasaxc) [reply npeps/npep-133-fqdn-egress-selector.md:134]: Calico's policy model only has one field and it feels quite natural in use so that'd be our preference.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (npinaeva) [reply npeps/npep-133-fqdn-egress-selector.md:306]: I love these stories :D
so did the problem happen because DNS matching happened after service loadbalancing? IOW, would matching network policy before service resolution solve this problem?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (smuda) [reply npeps/npep-133-fqdn-egress-selector.md:88]: Perhaps our user story (which really is a variant of the first user story) would clear anything up?

As a cluster admin, I want to allow certain pods in a namespace to access
an external service specified by a well-known domain name. The service is hosted on a cloud and changes IP-addresses often.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (fasaxc) [reply npeps/npep-133-fqdn-egress-selector.md:306]: Perhaps, but then it'd be very hard to interleave normal and DNS policy in anything that resembles iptables.  The DNS policy would happen "first" in order to see the pre-DNAT IPs but what if a higher priority normal policy should have dropped that traffic? BPF dataplanes that do policy and NAT can keep the pre-DNAT information around to match on at the policy stage.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (fasaxc) [inline npeps/npep-133-fqdn-egress-selector.md:141]: I asked my colleagues in our solutions team for examples of matching multiple URL segments, we do have customers that use * to match multiple parts in a prefix.  some examples they were able to share:

* `<team>.<platform>.example.com` matched with `*.example.com`
* `<sessionID>.<session>.<random>.<region>.cache.amazonaws.com` matched with `*.cache.amazonaws.com`

I think the AWS one would be `*.*.*.*.cache.amazonaws.com` in the proposed spec, which feels a bit messy.

Comment (npinaeva) [reply npeps/npep-133-fqdn-egress-selector.md:141]: very nice examples, thank you for gathering!
I am getting more convinced every day that `*` should match any number of labels, unless someone can point out a security/performance/etc concerns about doing that

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:134]: We discussed this last meeting, and the plan of record is to only support pattern, not name.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:133]: Updated the regex. Thanks for the suggestions!

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:114]: ack, switched to `domains`.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:141]: I'm ok with that -- I can see CDNs getting pretty deeply nested so having `*` match multiple labels should be ok.

I don't imagine this should have any performance implications. Security-wise, you sacrifice some explicitness. But on the other hand, when you allow `*.domain.com`, you've already indicated you trust the owner `domain.com`

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:154]: Sounds good to me. For the most part I believe this is compatible with what we're proposing.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:88]: @smuda I agree, I think this is well captured by the first user story

@astoycos PTAL, I re-worded a little bit, to  hopefully make it easier to understand.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (astoycos) [inline npeps/npep-133-fqdn-egress-selector.md:138]: In the meeting @npinaeva Proposed `DomainName` here however regardless Antonio suggested we always reference an RFC for decisions like this

Comment (fasaxc) [reply npeps/npep-133-fqdn-egress-selector.md:138]: https://datatracker.ietf.org/doc/html/rfc1034 S3.1 is the relevant one; 

> A domain is identified by a domain name, and consists of that part of
> the domain name space that is at or below the domain name which
> specifies the domain.

So a _Domain_ is a subtree of domain space whereas a _Domain Name_ is the textual name of one particular node in that tree.  Not sure that gives a clear answer to our naming dilemma though!  `*` patterns represent subtrees, FQDNs represent single nodes in the tree! Perhaps `DomainNamePatterns` conveys exactly what our field represents (but I think `Domains` and `DomainNames` are both defensible/attackable depending on how closely you read the RFC :laughing: )?

The "." suffix is also explained in that section.  A "." at the end of the domain name means it is absolute, whereas no "." means that it is relative to the local domain and the DNS resolver should add suffixes as needed.

Comment (npinaeva) [reply npeps/npep-133-fqdn-egress-selector.md:138]: I think a domain name with `*` is still considered a domain name. I read RFC as "domain name is a string, domain is an internal structure of the domain name space".
If we read closer to the wildcards definition S4.3.3, it says

> In the previous algorithm, special treatment was given to RRs with owner
names starting with the label "*".

S3.6 defines

> When we talk about a specific RR, we assume it has the following:
owner     -      which is the domain name where the RR is found.

So owner name is the domain name, and owner name may start with `*`, which makes it a valid domain name?
Reading RFCs is fun :D

File: npeps/npep-133-fqdn-egress-selector.md
Comment (aojea) [inline npeps/npep-133-fqdn-egress-selector.md:258]: does it mean that if implementation X considers kube-dns authoritative but the Pod uses a different nameserver the policy will not apply?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (aojea) [inline npeps/npep-133-fqdn-egress-selector.md:254]: I kind of infer that this is based on some implementation details, but is hard to get from the text why this will happen ...  "ability of workloads to resolve domains" means to filter DNS traffic or DNS packets? why should the policy filter things are not supposed to do? how is this different than any other traffic?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (aojea) [inline npeps/npep-133-fqdn-egress-selector.md:263]: is this implementable? this means the implementation has to cache all possible IPs of all the possible domains, no?

File: npeps/npep-133-fqdn-egress-selector.md
Comment (aojea) [inline npeps/npep-133-fqdn-egress-selector.md:267]: what TTL records is referring here? the TTL records of the DNS response that the Pod has received?
does it imply always the implementation to snoop?

Comment (aojea) [reply npeps/npep-133-fqdn-egress-selector.md:267]: ok, I see the table below, good work with that table, really intuitive

File: npeps/npep-133-fqdn-egress-selector.md
Comment (aojea) [inline npeps/npep-133-fqdn-egress-selector.md:333]: Other alternative is implement DNS filtering,  as cisco umbrella https://docs.umbrella.com/deployment-umbrella/docs/point-your-dns-to-cisco or  dnsfilter do ...

File: npeps/npep-133-fqdn-egress-selector.md
Comment (illrill) [reply npeps/npep-133-fqdn-egress-selector.md]: In https://github.com/cilium/cilium/issues/22081, I've been discussing this topic with maintainers of Cilium (wich regards to `CiliumNetworkPolicy`) . Though still not decided or implemented, the dialog has been leaning towards:

  - `**` matches any character (including `.`). Allows wildcarding of multiple levels of subdomains.
  - `*` matches any character except `.`. Allows wildcarding of a single subdomain.

Perhaps this can be useful, if not just for some additional perspectives.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (fasaxc) [reply npeps/npep-133-fqdn-egress-selector.md:138]: They do say that the owner is a domain(!) but they also say 
```
If any match, copy them into the answer
section, but set the owner of the RR to be QNAME, and
not the node with the "*" label.
```
I.e. don't send `*` on the wire and 
```
The owner name of the wildcard RRs is of
the form "*.<anydomain>", where <anydomain> is any domain name.
<anydomain> should not contain other * labels
```
Which implicitly says that a wildcard is not a domain so, it's a bit contradictory!

I think the wildcard comes from implementation concerns, they're really talking about the configuration for a DN server (and probably a particular one at that) and saying "you'll need wildcards in your implementation to handle mail and they should work like this to avoid these problems".

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:254]: I'm trying to clarify that this spec is explicitly **not** doing DNS filtering. So even though the policy says "allow traffic to kubernetes.io", you can still resolve any domains you like.

I can add a clarification explicitly saying that FQDN policies will not perform DNS Filtering.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:258]: As long as the two nameservers return the same IPs for domains it's not a problem. But if kube-dns says the IP is `1.2.3.4`, the implementation only allows `1.2.3.4`. If the pod's nameserver then says `5.6.7.8`, then traffic will be denied.

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:120]: Thanks for the great suggestion. Done!

File: npeps/npep-133-fqdn-egress-selector.md
Comment (rahulkjoshi) [reply npeps/npep-133-fqdn-egress-selector.md:274]: Yes we've pretty much agreed that the only way to correctly implement this is via a proxy that listens to DNS responses.

**Pull request merged**
